olive/passes/pytorch/quarot.py

```diff
@@ -0,0 +1,299 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+# -------------------------------------------------------------------------
+import logging
+import sys
+from enum import Enum
+from typing import Any, Dict, Union
+
+import torch
+import transformers
+from torch.utils.data import DataLoader, SubsetRandomSampler
+
+from olive.common.config_utils import validate_config
+from olive.constants import ModelFileFormat
+from olive.data.config import DataConfig
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import HfModelHandler, PyTorchModelHandler
+from olive.passes import Pass
+from olive.passes.olive_pass import PassConfigParam
+from olive.passes.pytorch.common import inherit_pytorch_from_hf
+
+logger = logging.getLogger(__name__)
+
+
+class QuaRot(Pass):
+    """A new Quantization scheme based on Rotations, which is able to quantize LLMs end-to-end.
+
+    See https://arxiv.org/pdf/2404.00456 for more details on the algorithm.
+
+    This pass only supports HfModelHandler.
+    """
+
+    class ModelDtype(str, Enum):
+        # input model's data type, we can assume the model is all float type
+        # sometime, the model is in double type, but we can convert it to float type
+        # before quantization
+        FP32 = "fp32"
+        FP16 = "fp16"
+        FP64 = "fp64"
+
+        def __str__(self) -> str:
+            return self.value
+
+        def get_torch_dtype(self):
+            return {
+                QuaRot.ModelDtype.FP32: torch.float32,
+                QuaRot.ModelDtype.FP16: torch.float16,
+                QuaRot.ModelDtype.FP64: torch.float64,
+            }[self]
+
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "input_model_dtype": PassConfigParam(
+                type_=QuaRot.ModelDtype,
+                default_value=QuaRot.ModelDtype.FP16,
+                description="Input model's data type.",
+            ),
+            "calibration_data_config": PassConfigParam(
```

Comment:
Hitesh and I had a related discussion before. Would it be possible to add the functionality for sampling directly to our dataloader component so that this and slicegpt pass don't have to expose `calibration_nsamples` and `calibration_batch_size`? 
wrt the `"labels"` as inputs, we could just wrap the original dataloader and override its iteration logic to return a single dict.

Reply:
Sounds good. How about implement it in follow-up PR?

Reply:
Sure! Follow up is good too.